# 用牛顿法求下列方程的根
- $x^2-e^x=0$
- $xe^x-1=0$
- $\lg x+x-2=0$

In [24]:
import math
def f1(x):
    if (2*x-math.exp(x) == 0):
        return "error"
    else:
        return x-(x*x-math.exp(x))/(2*x-math.exp(x))

def f2(x):
    if ((x+1)*math.exp(x) == 0):
        return "error"
    else:
        return x-(x*math.exp(x)-1)/((x+1)*math.exp(x))

def f3(x):
    if (x <= 0):
        return "error"
    if (1/(x*math.log(10))+1 == 0):
        return "error"
    else:
        return x-(math.log(x)/math.log(10)+x-2)/(1/(x*math.log(10))+1)

In [26]:
f3(1)

1.6972068934358862

In [30]:
x = 1
for i in range(10):
    x = f1(x)
    print(x)

-1.3922111911773332
-0.8350875293671394
-0.7098340945745987
-0.7034834042362847
-0.703467422599462
-0.7034674224983917
-0.7034674224983917
-0.7034674224983917
-0.7034674224983917
-0.7034674224983917


In [31]:
x = 1
for i in range(10):
    x = f2(x)
    print(x)

0.6839397205857212
0.5774544771544498
0.5672297377301171
0.5671432965302959
0.567143290409784
0.5671432904097838
0.5671432904097838
0.5671432904097838
0.5671432904097838
0.5671432904097838


In [32]:
x = 1
for i in range(10):
    x = f3(x)
    print(x)

1.6972068934358862
1.7553795434500208
1.75557949700258
1.755579499261178
1.755579499261178
1.755579499261178
1.755579499261178
1.755579499261178
1.755579499261178
1.755579499261178


## Result
- $x^2-e^x=0$ 根为 $-0.7034674224983917$
- $xe^x-1=0$ 根为 $0.5671432904097838$
- $\lg x+x-2=0$ 根为 $1.755579499261178$

# 编写一个割线法的程序，求解上述各方程
- $x^2-e^x=0$
- $xe^x-1=0$
- $\lg x+x-2=0$

In [33]:
import math
def f4(x1,x2):
    if ((x1*x1-math.exp(x1))-(x2*x2-math.exp(x2)) == 0):
        return x1
    else:
        return x1-(x1*x1-math.exp(x1))/((x1*x1-math.exp(x1))-(x2*x2-math.exp(x2)))*(x1-x2)

def f5(x1,x2):
    if ((x1*math.exp(x1)-1)-(x2*math.exp(x2)-1) == 0):
        return x1
    else:
        return x1-(x1*math.exp(x1)-1)/((x1*math.exp(x1)-1)-(x2*math.exp(x2)-1))*(x1-x2)

def f6(x1,x2):
    if (x1 <= 0 or x2 <= 0):
        return "error"
    if ((math.log(x1)/math.log(10)+x1-2)-(math.log(x2)/math.log(10)+x2-2) == 0):
        return x1
    else:
        return x1-(math.log(x1)/math.log(10)+x1-2)/((math.log(x1)/math.log(10)+x1-2)-(math.log(x2)/math.log(10)+x2-2))*(x1-x2)

In [34]:
x1 = 1
x2 = 0
for i in range(10):
    temp = x1
    x1 = f4(x1,x2)
    x2 = temp
    print(x1)

-1.3922111911773332
-0.20612751271406604
-0.5778333714634147
-0.7330842972939767
-0.7019544393346077
-0.7034498327184769
-0.7034674330341212
-0.7034674224983183
-0.7034674224983917
-0.7034674224983917


In [35]:
x1 = 1
x2 = 0
for i in range(10):
    temp = x1
    x1 = f5(x1,x2)
    x2 = temp
    print(x1)

0.36787944117144233
0.5033143321329856
0.5786158630519874
0.5665323438586994
0.5671375717285394
0.5671432932720224
0.5671432904097705
0.5671432904097838
0.5671432904097838
0.5671432904097838


In [ ]:
x1 = 1
x2 = 0.5
for i in range(10):
    temp = x1
    x1 = f6(x1,x2)
    x2 = temp
    print(x1)